python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -35,16 +46,186 @@ def __init__(self, params, lr):
+            }
+            for name, state in emb_state["states"].items():
+                if name not in name_to_index:
+                    raise ValueError(
+                        "loaded state dict contains a state {name}"
+                        "that can't be found in the optimizer states"
+                    )
+                state_idx = name_to_index[name]
+                state = state.to(states[state_idx].dtype)
+                state = state.to(th.device("cpu"))
+                states[state_idx][idx] = state
+        self._defaults.update(local_state_dict["params"])
+        self.__dict__.update(local_state_dict["params"])
+
+    def save(self, f):
+        """Save the local state_dict to disk on per rank.
+
+        Saved dict contains three parts:
+
+        * 'ids': list of global ids of the nodes/edges selected locally
+        * 'params': hyper parameters of the optimizer
+        * 'states': partial optimizer states indexed by ```ids``` of entire states
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, os.PathLike]
+            The path of the file.
+
+        See Also
+        --------
+        load
+        """
+        if self._world_size > 1:
+            th.distributed.barrier()
+        f = f if isinstance(f, str) else str(f, "UTF-8")
+        f = f"{f}_{self._rank}"
+        th.save(self.local_state_dict(), f)
+        if self._world_size > 1:
+            th.distributed.barrier()
+
+    def load(self, f):
+        """Load the local state of the optimizer from the file on per rank.
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, os.PathLike]
+            The path of the file.
+
+        See Also
+        --------
+        save
+        """
+        if self._world_size > 1:
+            th.distributed.barrier()
+        f = f if isinstance(f, str) else str(f, "UTF-8")
+        f_attach_rank = f"{f}_{self._rank}"
+        # Don't throw error here to support device number scale-out
+        # after reloading, but make sure your hyper parameter is same
+        # as before because new added local optimizers will be filled
+        # in nothing
+        if not exists(f_attach_rank):
+            warnings.warn(
+                f"File {f_attach_rank} can't be found, load nothing."
+            )
+        else:
+            old_world_size = self._load_state_from(f_attach_rank)
+            # Device number scale-in
+            if self._world_size < old_world_size:
+                for rank in range(
+                    self._rank + self._world_size, old_world_size, self._world_size
+                ):
+                    self._load_state_from(f"{f}_{rank}")
+        if self._world_size > 1:
+            th.distributed.barrier()
+
+    def _load_state_from(self, f):
+        local_state_dict = th.load(f)
+        world_size = local_state_dict["params"]["world_size"]
```

Comment:
merge these 2 lines into 1: `world_size = xxx_dict.pop('world_size') ?

Reply:
cool changed to pop